### PR TITLE
Change voucher types 

### LINKED ISSUE #4317: Change voucher types for new dashboard layout
We need to change voucher types for the new dashboard layout.

### PR DESCRIPTION
**Pull request by fowczarek**
I want to merge this change because, resolve #4317

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [ ] Database queries are optimized and the number of queries is constant.
1. [ ] Database migration files are up to date.
1. [ ] The changes are tested.
1. [ ] The code is documented (docstrings, project documentation).
1. [ ] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

File: tests/api/test_discount.py [hunk 1]
```diff
@@ -3,11 +3,11 @@
 import graphene
 import pytest
 from django_countries import countries
-from tests.api.utils import get_graphql_content
 
 from saleor.discount import DiscountValueType, VoucherType
 from saleor.discount.models import Sale, Voucher
 from saleor.graphql.discount.enums import DiscountValueTypeEnum, VoucherTypeEnum
+from tests.api.utils import get_graphql_content
```
Comment (NyanKiyoshi) [inline tests/api/test_discount.py]: Absolute import

**Comment by fowczarek** [comment]
> Are you planning on adding a migration?

This pull request has "In progress" label. I adding migrations after final rebase.

**Comment by django-queries** [comment]
Here is the report for b6cd11e52a9568a2f49707d11715d31157afc9d6 (mirumee/saleor @ 4317/change_voucher_types)
Base comparison is 482ba001bf3cbb8a6e765649e546b33bb58d00f7.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  add billing address to checkout     	         41	         41
  add shipping to checkout            	          7	          7
  checkout payment charge             	         16	         16
  complete checkout                   	          6	          6
  create checkout                     	         45	         45

# api.benchmark homepage
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve main menu                  	          5	          5
  retrieve product list               	          4	          4
  retrieve secondary menu             	          5	          5
  retrieve shop                       	          2	          2

# api.benchmark product
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  product details                     	         13	         13

# api.benchmark variant
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve variant list               	          9	          9
```

</p>
</details>

File: saleor/dashboard/discount/forms.py [hunk 2]
```diff
@@ -159,6 +159,38 @@ class CommonVoucherForm(forms.ModelForm):
     )
 
 
+class SpecificProductVoucherForm(CommonVoucherForm):
+    products = AjaxSelect2MultipleChoiceField(
+        queryset=Product.objects.all(),
+        fetch_data_url=reverse_lazy("dashboard:ajax-products"),
+        required=True,
+        label=pgettext_lazy("Product", "Products"),
```
Comment (NyanKiyoshi) [inline saleor/dashboard/discount/forms.py]: The context the doesn't tell the translators anything about what it is for

File: saleor/dashboard/discount/forms.py [hunk 3]
```diff
@@ -159,6 +159,38 @@ class CommonVoucherForm(forms.ModelForm):
...
+    )
+    categories = TreeNodeMultipleChoiceField(
+        queryset=Category.objects.all(),
+        required=True,
+        label=pgettext_lazy("Categories", "Categories"),
```
Comment (NyanKiyoshi) [inline saleor/dashboard/discount/forms.py]: Same as above

File: saleor/dashboard/discount/forms.py [hunk 4]
```diff
@@ -159,6 +159,38 @@ class CommonVoucherForm(forms.ModelForm):
...
+    class Meta:
+        model = Voucher
+        fields = [
+            "products",
+            "collections",
+            "categories",
+            "apply_once_per_order",
+            "min_amount_spent",
+        ]
+        labels = {"collections": pgettext_lazy("Collections", "Collections")}
```
Comment (NyanKiyoshi) [inline saleor/dashboard/discount/forms.py]: Same as above

File: saleor/dashboard/discount/forms.py [hunk 5]
```diff
@@ -159,6 +159,38 @@ class CommonVoucherForm(forms.ModelForm):
...
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.instance.pk:
+            self.fields["products"].set_initial(self.instance.products.all())
```
Comment (korycins) [inline saleor/dashboard/discount/forms.py]: Why do you init only `products`? What about categories and collections?
Is it expected?

File: saleor/dashboard/discount/forms.py [hunk 6]
```diff
@@ -159,6 +159,38 @@ class CommonVoucherForm(forms.ModelForm):
...
+        self.fields["categories"].required = False
+        self.fields["products"].required = False
```
Comment (korycins) [inline saleor/dashboard/discount/forms.py]: You miss `collections` here. Is it expected?

File: saleor/discount/__init__.py [hunk 7]
```diff
@@ -34,6 +35,12 @@ class VoucherType:
             pgettext_lazy("Voucher: discount for", "Specific categories of products"),
         ),
         (SHIPPING, pgettext_lazy("Voucher: discount for", "Shipping")),
+        (
+            SPECIFIC_PRODUCT,
+            pgettext_lazy(
+                "Voucher: discount for", "Specific products, collections and categories"
```
Comment (korycins) [inline saleor/discount/__init__.py]: `collections and categories` -> `collections or categories` ?

File: saleor/graphql/discount/mutations.py [hunk 8]
```diff
@@ -66,7 +66,11 @@ def remove_catalogues_from_node(cls, node, input):
 
 class VoucherInput(graphene.InputObjectType):
     type = VoucherTypeEnum(
-        description="Voucher type: product, category shipping or value."
+        description=(
+            "Voucher type: product, category shipping or entire order. "
```
Comment (korycins) [inline saleor/graphql/discount/mutations.py]: Types should be camelCase

File: saleor/graphql/discount/mutations.py [hunk 9]
```diff
@@ -66,7 +66,11 @@ def remove_catalogues_from_node(cls, node, input):
...
+            "Deprecated: PRODUCT, COLLECTION, CATEGORY are deprecated, "
```
Comment (korycins) [inline saleor/graphql/discount/mutations.py]: Here you have `deprecated` x2
Maybe something like: `"Deprecated fields: PRODUCT, COLLECTION, CATEGORY, use SPECIFIC_PRODUCT instead.`

File: saleor/graphql/schema.graphql [hunk 10]
```diff
@@ -3203,19 +3203,21 @@ type VoucherTranslation implements Node {
 }
 
 enum VoucherType {
-  VALUE
```
Comment (korycins) [inline saleor/graphql/schema.graphql]: I am not sure that you should remove this field. It should be marked as a deprecated also. As you did with PRODUCT, COLLECTION, CATEGORY

Comment (korycins) [reply saleor/graphql/schema.graphql]: If someone has own version of dashboard then it will be incompatible with these changes.

**Pull request merged**